tools/code-analyzer/src/commands/analyzer/index.ts

```diff
@@ -0,0 +1,470 @@
+					);
+				}
+			}
+		} else {
+			this.log(`\n## ${title}:`);
+			for (const [key, value] of data) {
+				this.log('FILE: ' + key);
+				this.log('---------------------------------------------------');
+				for (const [k, v] of value) {
+					this.log('HOOK: ' + k);
+					this.log(
+						'---------------------------------------------------'
+					);
+					this.log(` ${v[0].toUpperCase()} | ${v[1]} | ${v[2]}`);
+					this.log(
+						'---------------------------------------------------'
+					);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Scan patches for changes in templates
+	 *
+	 * @param {string} content Patch content.
+	 * @param {string} version Current product version.
+	 * @return {Promise<Map<string, string[]>>}
+	 */
+	private async scanTemplates(
+		content: string,
+		version: string
+	): Promise<Map<string, string[]>> {
+		CliUx.ux.action.start('Scanning template changes');
+
+		let report: Map<string, string[]> = new Map<string, string[]>();
+
+		if (!content.match(/diff --git a\/(.+)\/templates\/(.+)/g)) {
+			CliUx.ux.action.stop();
+			return report;
+		}
+
+		const matchPatches = /^a\/(.+)\/templates\/(.+)/g;
+		const title = 'Template change detected';
+		const patches = await this.getPatches(content, matchPatches);
+		const matchVersion = `^(\\+.+\\*.+)(@version)\\s+(${version.replace(
+			/\./g,
+			'\\.'
+		)}).*`;
+		const versionRegex = new RegExp(matchVersion, 'g');
+
+		for (const p in patches) {
+			const patch = patches[p];
+			const lines = patch.split('\n');
+			const filepath = await this.getFilename(lines[0]);
+			let code = 'warning';
+			let message = 'This template may require a version bump!';
+
+			for (const l in lines) {
+				const line = lines[l];
+
+				if (line.match(versionRegex)) {
+					code = 'notice';
+					message = 'Version bump found';
+				}
+			}
+
+			if ('notice' === code && report.get(filepath)) {
+				report.set(filepath, [code, title, message]);
+			} else if (!report.get(filepath)) {
+				report.set(filepath, [code, title, message]);
+			}
+		}
+
+		CliUx.ux.action.stop();
+		return report;
+	}
+
+	/**
+	 * Scan patches for hooks
+	 *
+	 * @param {string} content Patch content.
+	 * @param {string} version Current product version.
+	 * @return {Promise<Map<string, Map<string, string[]>>>}
+	 */
+	private async scanHooks(
+		content: string,
+		version: string
+	): Promise<Map<string, Map<string, string[]>>> {
+		CliUx.ux.action.start('Scanning for new hooks');
+
+		let report: Map<string, Map<string, string[]>> = new Map<
+			string,
+			Map<string, string[]>
+		>();
+
+		if (!content.match(/diff --git a\/(.+).php/g)) {
+			CliUx.ux.action.stop();
+			return report;
+		}
+
+		const matchPatches = /^a\/(.+).php/g;
+		const patches = await this.getPatches(content, matchPatches);
+		const verRegEx = await this.getVersionRegex(version);
+		const matchHooks = `@since\\s+(${verRegEx})(.*?)(apply_filters|do_action)\\(\\s+(\\'|\\")(.*?)(\\'|\\")`;
```

Comment:
Seems this pattern is incomplete. Missing a closing `)`. Also perhaps we should match `do_action( 'hook' );` and without space `do_action('hook');`.